### PR TITLE
Generalise hoistFreeT to Functor m

### PR DESCRIPTION
**Pull request by tomjaguarpaw**
Seems like this would be a worthy change.

I can't imagine why it could be wrong, but if so the fact that I made this PR is good evidence that its wrongness needs to be clearly documented!

### CODE REVIEW

File: src/Control/Monad/Trans/Free.hs [hunk 1]
```diff
@@ -422,8 +422,8 @@ instance (Monad m, Traversable m, Traversable f) => Traversable (FreeT f m) wher
 -- | Lift a monad homomorphism from @m@ to @n@ into a monad homomorphism from @'FreeT' f m@ to @'FreeT' f n@
 --
 -- @'hoistFreeT' :: ('Monad' m, 'Functor' f) => (m ~> n) -> 'FreeT' f m ~> 'FreeT' f n@
```
Comment (RyanGlScott) [inline src/Control/Monad/Trans/Free.hs:424]:
```suggestion
-- @'hoistFreeT' :: ('Functor' m, 'Functor' f) => (m ~> n) -> 'FreeT' f m ~> 'FreeT' f n@
```

File: src/Control/Monad/Trans/Free.hs [hunk 2]
```diff
@@ -422,8 +422,8 @@ instance (Monad m, Traversable m, Traversable f) => Traversable (FreeT f m) wher
...
-hoistFreeT :: (Monad m, Functor f) => (forall a. m a -> n a) -> FreeT f m b -> FreeT f n b
-hoistFreeT mh = FreeT . mh . liftM (fmap (hoistFreeT mh)) . runFreeT
+hoistFreeT :: (Functor m, Functor f) => (forall a. m a -> n a) -> FreeT f m b -> FreeT f n b
+hoistFreeT mh = FreeT . mh . fmap (fmap (hoistFreeT mh)) . runFreeT
```
Comment (RyanGlScott) [inline src/Control/Monad/Trans/Free.hs:426]: `hoistFreeT` in `Control.Monad.Trans.Free.Ap` could also benefit from replacing `liftM` with `fmap`:

https://github.com/ekmett/free/blob/5721f23f711cebaa93bc6292ed025a80fbc4f06b/src/Control/Monad/Trans/Free/Ap.hs#L422-L423

**Review by RyanGlScott — COMMENTED** [review]
This looks reasonable to me. There are some build errors on old versions of GHC where `Functor` is not a superclass of `Monad`, but those should be fixable by just adding `Functor` constraints to the affected places.

File: src/Control/Monad/Trans/Free.hs (hunk 2 above)
Comment (tomjaguarpaw) [reply src/Control/Monad/Trans/Free.hs:426]: OK, I have incorporated that.

**Comment by tomjaguarpaw** [comment]
This one should pass (at least it passed in the Action on my fork).

File: src/Control/Monad/Trans/Free/Ap.hs [hunk 3]
```diff
@@ -419,8 +419,8 @@ instance (Monad m, Traversable m, Traversable f) => Traversable (FreeT f m) wher
 -- | Lift a monad homomorphism from @m@ to @n@ into a monad homomorphism from @'FreeT' f m@ to @'FreeT' f n@
 --
 -- @'hoistFreeT' :: ('Monad' m, 'Functor' f) => (m ~> n) -> 'FreeT' f m ~> 'FreeT' f n@
```
Comment (RyanGlScott) [inline src/Control/Monad/Trans/Free/Ap.hs:421]:
```suggestion
-- @'hoistFreeT' :: ('Functor' m, 'Applicative' f) => (m ~> n) -> 'FreeT' f m ~> 'FreeT' f n@
```

**Review by RyanGlScott — COMMENTED** [review]
LGTM, aside from one minor Haddock suggestion.

**Comment by tomjaguarpaw** [comment]
Ah, thanks for spotting that. Applied.

**Review by RyanGlScott — APPROVED** [review]

**Pull request merged**